proto-lens-protoc/src/Data/ProtoLens/Compiler/Generate.hs

```diff
@@ -259,80 +259,185 @@ generateEnumExports syntaxType e = [exportAll n, exportWith n aliases]
 
 generateEnumDecls :: SyntaxType -> EnumInfo Name -> [Decl]
 generateEnumDecls Proto3 info =
-    -- newtype Foo = Foo Int32
-    [ newtypeDecl dataName "Data.Int.Int32"
-        $ deriving' ["Prelude.Eq", "Prelude.Ord", "Prelude.Enum", "Prelude.Bounded"]
-
-    -- instance Show Foo where
-    --    showsPrec _ Value0 = "Value0" -- the Haskell name
-    --    showsPrec p (Foo k) = showParen (p > 10)
-    --                            $ showString "toEnum " . shows k
-    , instDecl [] ("Prelude.Show" `ihApp` [dataType])
-        [ [ match "showsPrec" [pWildCard, pApp (unQual n) []]
-               $ "Prelude.showString" @@ stringExp (prettyPrint n)
-          | n <- map enumValueName $ enumValues info
-          ]
-          ++
-          [ match "showsPrec" ["p", pApp (unQual dataName) ["k"]]
-                $ "Prelude.showParen" @@ ("Prelude.>" @@ "p" @@ litInt 10)
-                          @@ ("Prelude.." @@ ("Prelude.showString"
-                                                  @@ stringExp "toEnum ")
-                                          @@ ("Prelude.shows" @@ "k"))
-          ]
-        ]
+    -- data FooEnum = Enum1
+    --   | Enum2
+    --   | FooEnum'Unrecognized !FooEnum'UnrecognizedValue
+    --   deriving (Prelude.Show, Prelude.Eq, Prelude.Ord, Prelude.Read)
+    [ dataDecl dataName
+        (  (flip conDecl [] <$> constructorNames)
+        ++ [conDecl unrecognizedName [unrecognizedValueType]]
+        )
+        $ deriving' ["Prelude.Show", "Prelude.Eq", "Prelude.Ord, Prelude.Read"]
 
-    -- instance MessageEnum Foo where
-    --    maybeToEnum k = Just $ toEnum k
-    --    showEnum (Foo 0) = "Value0" -- the proto name
-    --    showEnum (Foo k) = show k
-    --    readEnum "Value0" = Just (Foo 0)
-    --    readEnum _ = Nothing
+    -- newtype FooEnum'UnrecognizedValue = FooEnum'UnrecognizedValue Data.Int.Int32
+    --   deriving (Prelude.Eq, Prelude.Ord, Prelude.Show, Prelude.Read)
+    , newtypeDecl unrecognizedValueName
+       "Data.Int.Int32"
+        $ deriving' ["Prelude.Eq", "Prelude.Ord", "Prelude.Show, Prelude.Read"]
+
+    -- instance Data.ProtoLens.MessageEnum FooEnum where
+    --       maybeToEnum 0 = Prelude.Just Enum1
+    --       maybeToEnum 3 = Prelude.Just Enum2
+    --       maybeToEnum k
+    --         = Prelude.Just
+    --             (FooEnum'Unrecognized
+    --               (FooEnum'UnrecognizedValue (Prelude.fromIntegral k)))
+    --       showEnum (FooEnum'Unrecognized (FooEnum'UnrecognizedValue k))
+    --         = Prelude.show k
+    --       showEnum k = Prelude.show k
+    --       readEnum "Enum2a" = Prelude.Just Enum2a -- alias
+    --       readEnum k = Text.Read.readMaybe k
```

Comment:
`readMaybe k >>= maybeToEnum` assumes that the textual representation of these enum's is an Int. I was trying to reproduce the pre-existing logic in which strings like "Enum2a", "Enum2", etc are converted into instances of the sum type

Reply:
Oh, I see, I think I misread your code earlier.

Taking another look, we actually can't use the `Show` and `Read` instances for `FooEnum` to implement `showEnum/readEnum`.  Because for nested enums like
```
message Bar {
  enum Foo { A = 1; }
}
```
we want `showEnum Bar'A = "A"` and `readEnum "A" = Bar'A`, not `showEnum Bar'A = "Bar'A"` and `readEnum "Bar'A" = Bar'A`.  It could probably be described better in the docs, but `readEnum/showEnum` are specifically for the proto text format representation of the enum:
https://github.com/google/proto-lens/blob/master/proto-lens/src/Data/ProtoLens/Message.hs#L231

I think we should just take the string representations (*not* the ints like I suggested earlier), as in the code for proto2:
https://github.com/google/proto-lens/blob/master/proto-lens-protoc/src/Data/ProtoLens/Compiler/Generate.hs#L374

```
readEnum "Enum2" = Enum2
readEnum "Enum2a" = Enum2a
etc.
```

Reply:
That sounds good, but that still leaves us with no ability to serialize/deserialize unrecognized enum's. I'm assuming `readEnum . showEnum == id` is a property we want to maintain. If so what should `showEnum Foo'FooEnum Foo'Unrecognized...` result in?

Reply:
Ah, good point, we do need something sensible for `showEnum` at least.  How about this?
```
showEnum Enum2 = "Enum2"
showEnum (Foo'Unrecognized (Foo'UnrecognizedValue n)) = show n

readEnum "Enum2" = Enum2
readEnum "Enum2a" = Enum2
readEnum s = Foo'Unrecognized . UnrecognizedValue <$> readMaybe s
```

Since in proto text format, it's fine to refer to an enum by its numeric value rather than the case name.